File: src/windows/Maple.WinUI/MapleUI/MoleculesL2/MuiBotOutput.cs
using System;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Maple.UI.Atoms;

namespace Maple.UI
{
    /// <summary>
    /// Maple.UI Bot Output molecule (unified-component-catalog.md §3,
    /// "Bot Output" row: "Streaming generated result", built from Text,
    /// Progress, Avatar) — an avatar plus generated text that, while
    /// <see cref="Streaming"/>, reveals progressively via a real WinUI
    /// <see cref="DispatcherTimer"/> ticking a plain, WinUI-free
    /// <see cref="MuiBotOutputTicker"/> — same "DispatcherTimer drives a
    /// testable plain-C# core" split <see cref="Atoms.MuiToastAutoDismissTimer"/>
    /// uses. Ports `mui-bot-output.component.ts`'s <c>ngOnChanges</c>
    /// restart-on-Text/Streaming-change contract exactly.
    /// </summary>
    public sealed class MuiBotOutput : ContentControl
    {
        public static readonly DependencyProperty TextProperty =
            DependencyProperty.Register(nameof(Text), typeof(string), typeof(MuiBotOutput),
                new PropertyMetadata(string.Empty, (d, _) => ((MuiBotOutput)d).Restart()));

        public static readonly DependencyProperty StreamingProperty =
            DependencyProperty.Register(nameof(Streaming), typeof(bool), typeof(MuiBotOutput),
                new PropertyMetadata(false, (d, _) => ((MuiBotOutput)d).Restart()));

        public static readonly DependencyProperty BotNameProperty =
            DependencyProperty.Register(nameof(BotName), typeof(string), typeof(MuiBotOutput),
                new PropertyMetadata("Maple AI", (d, e) => ((MuiBotOutput)d)._avatar.Name = (string)e.NewValue));

        public static readonly DependencyProperty CharsPerTickProperty =
            DependencyProperty.Register(nameof(CharsPerTick), typeof(int), typeof(MuiBotOutput),
                new PropertyMetadata(2, (d, _) => ((MuiBotOutput)d).Restart()));

        public static readonly DependencyProperty IntervalMsProperty =
            DependencyProperty.Register(nameof(IntervalMs), typeof(int), typeof(MuiBotOutput),
                new PropertyMetadata(30, (d, e) => ((MuiBotOutput)d)._timer.Interval = TimeSpan.FromMilliseconds((int)e.NewValue)));

        public string Text
        {
            get => (string)GetValue(TextProperty);
            set => SetValue(TextProperty, value);
        }

        /// <summary>When true, <see cref="Text"/> is revealed progressively
        /// rather than shown whole.</summary>
        public bool Streaming
        {
            get => (bool)GetValue(StreamingProperty);
            set => SetValue(StreamingProperty, value);
        }

        public string BotName
        {
            get => (string)GetValue(BotNameProperty);
            set => SetValue(BotNameProperty, value);
        }

        public int CharsPerTick
        {
            get => (int)GetValue(CharsPerTickProperty);
            set => SetValue(CharsPerTickProperty, value);
        }

        public int IntervalMs
        {
            get => (int)GetValue(IntervalMsProperty);
            set => SetValue(IntervalMsProperty, value);
        }

        /// <summary>Fires once when the full Text has been revealed.</summary>
        public event EventHandler? Completed;

        private readonly StackPanel _root = new() { Orientation = Orientation.Horizontal, Spacing = 10 };
        private readonly MuiAvatar _avatar = new() { AvatarSize = MuiAvatarSize.Sm, Name = "Maple AI" };
        private readonly StackPanel _content = new() { Orientation = Orientation.Vertical, Spacing = 6 };
        private readonly MuiText _text = new() { Variant = MuiTextVariant.Body, ColorRole = MuiTextColorRole.Muted };
        private readonly MuiProgress _ring = new() { ProgressShape = MuiProgressShape.Ring, ProgressSize = MuiProgressSize.Sm, IsIndeterminate = true, Label = "Generating" };
        private readonly DispatcherTimer _timer = new() { Interval = TimeSpan.FromMilliseconds(30) };

        private MuiBotOutputTicker _ticker = new(2);

        public MuiBotOutput()
        {
            _content.Children.Add(_text);
            _content.Children.Add(_ring);
            _root.Children.Add(_avatar);
            _root.Children.Add(_content);
            Content = _root;
            IsTabStop = false;

            _timer.Tick += (_, _) => OnTick();

            Restart();
        }

        private void OnTick()
        {
            _ticker.Tick((Text ?? string.Empty).Length);
            ApplyVisible();
            if (_ticker.VisibleLength >= (Text ?? string.Empty).Length)
            {
                _timer.Stop();
                Completed?.Invoke(this, EventArgs.Empty);
            }
        }

        private void Restart()
        {
            _timer.Stop();
            _ticker = new MuiBotOutputTicker(Math.Max(1, CharsPerTick));

            if (Streaming)
            {
                ApplyVisible();
                _timer.Start();
            }
            else
            {
                _ticker.RevealAll((Text ?? string.Empty).Length);
                ApplyVisible();
            }
        }

        private void ApplyVisible()
        {
            var text = Text ?? string.Empty;
            _text.Text = text.Substring(0, Math.Min(_ticker.VisibleLength, text.Length));
            _ring.Visibility = Streaming && _ticker.VisibleLength < text.Length ? Visibility.Visible : Visibility.Collapsed;
        }
    }
}

File: src/apple/Packages/MapleUI/Sources/MapleUI/MoleculesL2/MuiBotOutput.swift
// MuiBotOutput.swift — Maple UI Molecules-L2 (unified-component-
// catalog.md §3). Streaming generated result, built from Text, Progress,
// Avatar.
//
// Reveal timing is driven by an injected clock — same split
// `MuiToastController`/`MuiCodeBlockController` use for their own timing —
// so the tick loop is unit-testable without racing the real wall clock.

import SwiftUI

/// Drives the character-by-character reveal — factored out of the view so
/// it's unit-testable against an injected clock instead of the real timer
/// the web reference drives with `setInterval`.
@MainActor
final class MuiBotOutputController: ObservableObject {
    @Published private(set) var visibleLength = 0

    private let sleep: @Sendable (UInt64) async -> Void

    init(sleep: @escaping @Sendable (UInt64) async -> Void = { ns in try? await Task.sleep(nanoseconds: ns) }) {
        self.sleep = sleep
    }

    /// Reveals `text` progressively, `charsPerTick` characters every
    /// `intervalMs`, until the whole string is visible — or shows it
    /// whole immediately when `streaming` is false. Calls `onCompleted`
    /// exactly once, when the full text becomes visible under streaming
    /// (never for a non-streaming reveal — matching the web reference's
    /// `completed` output, which only fires from its timer path). Exits
    /// early without calling `onCompleted` if the enclosing task is
    /// cancelled (the view disappeared, or `text`/`streaming` changed
    /// mid-reveal).
    func reveal(text: String, streaming: Bool, charsPerTick: Int, intervalMs: Int, onCompleted: () -> Void) async {
        guard streaming else {
            visibleLength = text.count
            return
        }
        visibleLength = 0
        while visibleLength < text.count {
            if Task.isCancelled { return }
            await sleep(UInt64(max(1, intervalMs)) * 1_000_000)
            if Task.isCancelled { return }
            visibleLength = Swift.min(text.count, visibleLength + max(1, charsPerTick))
        }
        onCompleted()
    }
}

public struct MuiBotOutput: View {
    public let text: String
    /// When true, `text` is revealed progressively rather than shown whole.
    public let streaming: Bool
    public let botName: String
    /// Characters revealed per tick.
    public let charsPerTick: Int
    public let intervalMs: Int
    /// Fires once when the full `text` has been revealed under streaming.
    public let completed: (() -> Void)?

    @StateObject private var controller: MuiBotOutputController

    public init(
        text: String,
        streaming: Bool = false,
        botName: String = "Maple AI",
        charsPerTick: Int = 2,
        intervalMs: Int = 30,
        completed: (() -> Void)? = nil
    ) {
        self.text = text
        self.streaming = streaming
        self.botName = botName
        self.charsPerTick = charsPerTick
        self.intervalMs = intervalMs
        self.completed = completed
        self._controller = StateObject(wrappedValue: MuiBotOutputController())
    }

    init(
        text: String,
        streaming: Bool,
        botName: String,
        charsPerTick: Int,
        intervalMs: Int,
        completed: (() -> Void)?,
        controller: @autoclosure @escaping () -> MuiBotOutputController
    ) {
        self.text = text
        self.streaming = streaming
        self.botName = botName
        self.charsPerTick = charsPerTick
        self.intervalMs = intervalMs
        self.completed = completed
        self._controller = StateObject(wrappedValue: controller())
    }

    public var body: some View {
        HStack(alignment: .top, spacing: MuiTokens.spacingSm) {
            MuiAvatar(name: botName, size: .sm)
            VStack(alignment: .leading, spacing: MuiTokens.spacingXs) {
                MuiText(visibleText, variant: .body, color: .muted, block: true)
                if isRevealing {
                    MuiProgress(shape: .ring, size: .sm, value: nil, label: "Generating")
                }
            }
        }
        .task(id: revealKey) {
            await controller.reveal(text: text, streaming: streaming, charsPerTick: charsPerTick, intervalMs: intervalMs) {
                completed?()
            }
        }
        .accessibilityElement(children: .combine)
    }

    private var visibleText: String {
        String(text.prefix(controller.visibleLength))
    }

    private var isRevealing: Bool {
        streaming && controller.visibleLength < text.count
    }

    /// Re-runs the reveal task whenever the source text or streaming flag
    /// changes — mirrors the web reference's `ngOnChanges` gate on those
    /// two inputs.
    private var revealKey: String {
        "\(text)|\(streaming)"
    }
}

#Preview("MuiBotOutput") {
    VStack(alignment: .leading, spacing: 16) {
        MuiBotOutput(text: "This photo was likely taken at golden hour near a coastline.", streaming: true)
        MuiBotOutput(text: "Already-complete caption.", streaming: false)
    }
    .padding()
    .frame(width: 340)
    .background(MuiTokens.bg)
}
